File: 2022/03/13/minDepthBinaryTree.swift
// https://leetcode.com/problems/minimum-depth-of-binary-tree/
// Author: Andrew Jarombek
// Date: 3/13/2022

public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init() { self.val = 0; self.left = nil; self.right = nil; }
    public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
    public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
        self.val = val
        self.left = left
        self.right = right
    }
}

class Solution {
    func minDepth(_ root: TreeNode?) -> Int {
        if root == nil {
            return 0
        }

        return dfs(root!, 0)
    }

    func dfs(_ node: TreeNode, _ level: Int) -> Int {
        if node.left == nil && node.right == nil {
            return 1
        }

        var right = Int.max
        var left = Int.max

        if let rightNode = node.right {
            right = dfs(rightNode, level + 1)
        }

        if let leftNode = node.left {
            left = dfs(leftNode, level + 1)
        }

        return 1 + min(right, left)
    }
}

File: 2022/11/13/cpp/minDepthBinaryTree.cpp
// https://leetcode.com/problems/minimum-depth-of-binary-tree/
// Author: Andrew Jarombek
// Date: 11/13/2022

#include <cassert>
#include <algorithm>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

int minDepth(TreeNode* root);
int dfs(TreeNode* node, int depth);

int minDepth(TreeNode* root) {
    if (root == nullptr)
        return 0;

    return dfs(root, 1);
}

int dfs(TreeNode* node, int depth) {
    if (node->left == nullptr && node->right == nullptr) {
        return depth;
    }

    int minDepth = INT32_MAX;

    if (node->left != nullptr) {
        minDepth = dfs(node->left, depth + 1);
    }

    if (node->right != nullptr) {
        minDepth = min(dfs(node->right, depth + 1), minDepth);
    }

    return minDepth;
}

int main() {
    auto* tree = new TreeNode(1, new TreeNode(2), new TreeNode(3));
    assert(minDepth(tree) == 2);
}